File: Algorithms/p241_Different_Ways_to_Add_Parentheses/p241_Different_Ways_to_Add_Parentheses.py
class Solution(object):
    def diffWaysToCompute(self, input):
        """
        :type input: str
        :rtype: List[int]
        """
        operands = []
        operators = []
        while True:
            i = 0
            while i<len(input) and not input[i] in "+-*":
                i += 1
            operands.append(int(input[0:i]))
            input = input[i:]
            if not input:
                break
            operators.append(input[0])
            input = input[1:]
        return self.subExpressionResults(operands,operators,0,len(operands)-1)
        
    # i,j: subrange of operands
    def subExpressionResults(self, operands, operators, i, j):
        results = []
        if i<j:
            for k in range(i,j):
                results_l = self.subExpressionResults(operands,operators,i,k)
                results_r = self.subExpressionResults(operands,operators,k+1,j)
                for v1 in results_l:
                    for v2 in results_r:
                        if operators[k]=='+':
                            v = v1+v2
                        elif operators[k]=='-':
                            v = v1-v2
                        else:
                            v = v1*v2
                        results.append(v)
        else:
            results.append(operands[i])
        return results

File: Algorithms/p241_Different_Ways_to_Add_Parentheses/p241_Different_Ways_to_Add_Parentheses.cpp
class Solution {
public:
    vector<int> diffWaysToCompute(string input) {
        vector<int> operands;
        vector<int> operators;
        while(1){
            string::size_type end;
            operands.push_back(stoi(input,&end));
            input.erase(0,end);
            if(input.empty()){
                break;
            }
            if(input[0]=='+'){
                operators.push_back(0);
            }else if(input[0]=='-'){
                operators.push_back(1);
            }else{
                operators.push_back(2);
            }
            input.erase(0,1);
        }
        return subExpressionResults(operands,operators,0,operands.size()-1);
    }
    
    // operators[i]: 0:+, 1:-, 2:*
    // i,j: subrange of operands
    vector<int> subExpressionResults(const vector<int>& operands, const vector<int>& operators, int i, int j){
        vector<int> results;
        if(i<j){
            // k: operator index
            for(int k=i;k<j;++k){
                vector<int> results_l = subExpressionResults(operands,operators,i,k);
                vector<int> results_r = subExpressionResults(operands,operators,k+1,j);
                for(int v1 : results_l){
                    for(int v2 : results_r){
                        int v;
                        if(operators[k]==0){
                            v = v1+v2;
                        }else if(operators[k]==1){
                            v = v1-v2;
                        }else{
                            v = v1*v2;
                        }
                        results.push_back(v);
                    }
                }
            }
        }else{
            results.push_back(operands[i]);
        }
        return results;
    }
};
